src/libraries/System.Linq/src/System/Linq/CountBy.cs

```diff
@@ -24,22 +25,24 @@ public static partial class Enumerable
                 return [];
             }
 
-            return CountByIterator(source, keySelector, keyComparer);
+            return new CountByEnumerable<TSource, TKey>(source, keySelector, keyComparer);
         }
 
-        private static IEnumerable<KeyValuePair<TKey, int>> CountByIterator<TSource, TKey>(IEnumerable<TSource> source, Func<TSource, TKey> keySelector, IEqualityComparer<TKey>? keyComparer) where TKey : notnull
+        private sealed class CountByEnumerable<TSource, TKey>(IEnumerable<TSource> source, Func<TSource, TKey> keySelector, IEqualityComparer<TKey>? keyComparer) : IEnumerable<KeyValuePair<TKey, int>> where TKey : notnull
         {
-            using IEnumerator<TSource> enumerator = source.GetEnumerator();
-
-            if (!enumerator.MoveNext())
+            public IEnumerator<KeyValuePair<TKey, int>> GetEnumerator()
             {
-                yield break;
-            }
+                using IEnumerator<TSource> enumerator = source.GetEnumerator();
 
-            foreach (KeyValuePair<TKey, int> countBy in BuildCountDictionary(enumerator, keySelector, keyComparer))
-            {
-                yield return countBy;
+                if (!enumerator.MoveNext())
+                {
+                    return ((IEnumerable<KeyValuePair<TKey, int>>)[]).GetEnumerator();
+                }
+
+                return BuildCountDictionary(enumerator, keySelector, keyComparer).GetEnumerator();
```

Comment:
One issue I'm finding with this approach is that this is returning an enumerator whose implementation is public. Couldn't this result users taking a dependency on that implementation details which might prevent us from evolving the method in the future?

Reply:
A dictionary enumerator does not have any useful method or property, and a dictionary enumerator is hidden by the IEnumerator interface.

Reply:
Not sure I follow, wouldn't this cast work?
```C#
(Dictionary<T, int>.Enumerator)source.CountBy(x => x).GetEnumerator();
```

Reply:
> Not sure I follow, wouldn't this cast work?

Yes, it will work. But `Dictionary.Enumerator` does not have any specific method or property.

Reply:
Changing the implementation to not use dictionary would result in this cast failing for anybody who might depend on this. This is why we never return enumerable/enumerator implementations that are public unless they are specifically present in the type signature of the method (as is the case with `ToList()` or the inner array in `Chunk()`).

Reply:
I haven't seen a code analyzer that would check this rule, I doubt it is so, why isn't there a code analyzer for this rule?